### DISCUSSION TITLE
How to configure a freshness policy sensor on a high quantity of assets?

### DISCUSSION THREAD

**Discussion by marcilj** [dagster-io/dagster#19232] [Q&A] [answered]
Was trying to reproduce the configure describe here
https://docs.dagster.io/concepts/partitions-schedules-sensors/asset-sensors#freshness-policy-sensors

```python
from dagster import FreshnessPolicySensorContext, freshness_policy_sensor


@freshness_policy_sensor(asset_selection=AssetSelection.all())
def my_freshness_alerting_sensor(context: FreshnessPolicySensorContext):
    if context.minutes_overdue is None or context.previous_minutes_overdue is None:
        return

    if context.minutes_overdue >= 10 and context.previous_minutes_overdue < 10:
        send_alert(
            f"Asset with key {context.asset_key} is now more than 10 minutes overdue."
        )
    elif context.minutes_overdue == 0 and context.previous_minutes_overdue >= 10:
        send_alert(f"Asset with key {context.asset_key} is now on time.")
```

The problem I have is that the function is always timing out.
It's taking more than 1 minute to run and Dagster complains and fail the sensor run.

I know the feature is experimental, but even If I change the logic to create my own sensor and loop over assets freshness policies, I'll endup with the same timeout problem.

Is there a way to avoid getting timed out, like increasing the sensors timeout value? Or anyway I could run that sensor in a loop so it would timed out?

**Comment by jamiedemaria** [comment, marked as answer]
Hey @marcilj - so the cause of the issue is that the `@freshness_policy_sensor` creates a normal `@sensor` that runs the decorated function for each asset in the `asset_selection`. Since you have do many assets, the number of times the `my_freshness_alerting_sensor` function is called is timing out the wrapping sensor. One option would be to break up the sensor into multiple sensors that target smaller groups of assets. For example if you are already grouping your assets by some other organization scheme, then you could have a freshness sensor for each of those groups

**Reply by marcilj** [reply]
I tried to do that and it still failed... Seems like 66 assets in a group is to much...

Isn't there another way we can achieve that instead of sensors?

It feels weird to build the same sensors multiple times only because we have that time out limitation.

**Reply by jamiedemaria** [reply]
@OwenKephart  do you have any advice on this?

**Reply by marcilj** [reply]
Trying to bring back that thread if any solution exist currently.

**Reply by OwenKephart** [reply]
Hey @marcilj , unfortunately there's not a great existing solution here beyond splitting this into more sensors (as discussed above, and this may require some sort of factory pattern to break this up into a larger set of sensors), or raising the sensor timeout.

One way of increasing the timeout for your sensors is setting the `DAGSTER_SENSOR_GRPC_TIMEOUT_SECONDS` environment variable to a larger value in the process that executes your sensor code.
